torchinductor/codegen/autotuner.py

```diff
@@ -0,0 +1,166 @@
+    #     self.fn.set_args(*args, **kwargs)
+    #     # get candidate kernels for fn
+    #     self.kernels = self.fn.candidate_kernels()
+    #     # filter kernels that args/kwargs does not meet requirements
+    #     self.fn.filter_kernels(self.kernels)
+    #     # if only one choice, return that kernel
+    #     if len(self.kernels) == 1:
+    #         kernel = self.kernels[0]
+    #         return kernel(*args, **kwargs)
+    #     # use input shape, stride, layer parameters as key to cache the best kernel to use
+    #     key = self.fn.gen_key()
+    #     if key not in self.cache:
+    #         bench_start = time.time()
+    #         timings = {
+    #             kernel: self._bench(*args, kernel=kernel, **kwargs) for kernel in self.kernels
+    #         }
+    #         bench_end = time.time()
+    #         self.bench_time = bench_end - bench_start
+    #         self.cache[key] = builtins.min(timings, key=timings.get)
+    #         self.kernels_timings = timings
+    #         kernel = self.cache[key]
+    #     self.best_kernel = kernel
+    #     return self.best_kernel(*args, **kwargs)
+
+
+autotune = Autotuner()
+
+
+def tuned_conv(*args, **kwargs):
+
+    # gather info
+    x = args[0]
+    w = args[1]
+    # bias = args[2]
+    stride = args[3]
+    padding = args[4]
+    dilation = args[5]
+    transposed = args[6]
+    output_padding = args[7]
+    groups = args[8]
+
+    BATCH, IN_C, IN_H, IN_W = x.shape
+    KERNEL_N, _, KERNEL_H, KERNEL_W = w.shape
+    stride = stride
+    padding = padding
+    dilation = dilation
+    transposed = transposed
+    output_padding = output_padding
+    groups = groups
+    stride_x = x.stride()
+    stride_w = w.stride()
+    # the identifiable args for the layers
+    id_args = [
+        BATCH,
+        IN_C,
+        IN_H,
+        IN_W,
+        KERNEL_N,
+        KERNEL_H,
+        KERNEL_W,
+        stride,
+        padding,
+        dilation,
+        transposed,
+        output_padding,
+        groups,
+        stride_x,
+        stride_w,
+    ]
+    use_cuda = x.is_cuda
+
+    # gen_key
+    key = tuple([arg for arg in id_args])
+    key = ("conv",) + key
+
+    # candidate kernels
+    kernels = ["aten.convolution"]
+    if use_cuda:
+        kernels += ["triton_ops.conv", "triton_ops.conv1x1"]
+
+    # filter kernels that args/kwargs does not meet requirements
+    remove_kernels = []
+    if groups > 1 or transposed:
+        remove_kernels += ["triton_ops.conv", "triton_ops.conv1x1"]
+    # triton_ops.conv1x1 could only deal with nhwc 1x1 kernel
+    if stride_x[1] > 1 or KERNEL_H > 1 or KERNEL_W > 1:
+        remove_kernels += ["triton_ops.conv1x1"]
+    kernels = [k for k in kernels if k not in remove_kernels]
+
+    # str to callable functions
+    kernel_dict = {
+        "aten.convolution": aten.convolution,
+        "triton_ops.conv": triton_ops.conv,
+        "triton_ops.conv1x1": triton_ops.conv1x1,
```

Comment:
Is it always valid to use conv1x1, or do we need to check if it is a 1x1 conv?

Reply:
`kernels = [k for k in kernels if k not in remove_kernels]`. filter kernels that args/kwargs does not meet requirements of certain backend and put them into `remove_kernels`